src/ripple/nodestore/DeterministicShard.md

```diff
@@ -0,0 +1,135 @@
+as values of header field both for `nudb.key` and `nudb.dat` files.
+```
+Version         2
+UID             digest(0)
+Appnum          digest(2) | 0x5348524400000000 /* 'SHRD' */
+KeySize         32
+Salt            digest(1)
+Pepper          XXH64(Salt)
+BlockSize       0x1000 (4096 bytes)
+LoadFactor      0.5 (numeric 0x8000)
+```
+Note: XXH64() is well-known hash algorithm.
+
+The `digest(i)` mentioned above defined as the follows:
+
+First, RIPEMD160 hash `H` calculated of the following structure
+(the same as final Key of the shard):
+```
+uint32          version         Version of shard, 2 at the present
+uint32          firstSeq        Sequence number of first ledger in the shard
+uint32          lastSeq         Sequence number of last ledger in the shard
+uint256         lastHash        Hash of last ledger in shard
+```
+there all 32-bit integers are hashed in network byte order
+(bigendian: most significant byte first).
+
+Then, `digest(i)` is defined as the following part of the above hash `H`:
+```
+digest(0) = H[0] << 56 | H[1] << 48 | ... | H[7] << 0,
+digest(1) = H[8] << 56 | H[9] << 48 | ... | H[15] << 0,
+digest(2) = H[16] << 24 | H[17] << 16 | ... | H[19] << 0,
+```
+where `H[i]` denotes `i`-th byte of hash `H`.
+
+
+## Contents
+
+After deterministic shard is created using the above mentioned headers,
+it filled with objects using the following steps.
+
+1. All ledgers of the shard are visited in descendent order (from high
+ledger sequences to low).
+
+2. For each ledger all the objects in it are visited is some order
+(not matter what is this order). Here the objects are: ledgers, SHAmap tree
+nodes including accounts and transactions, final key. Final key object visited
+after last visited ledger.
+
+3. Set of ledgers within the shard divide into groups. First group
+consists of first k1 visited ledgers where k1 is the minimum number
+for which total count of visited objects in first k1 ledgers not less than
+16384. Second group consists of next k2 ledgers after first k1 ones where
+k2 is the minimum number for which total count of visited objects in these
+k2 ledgers not less than 16384. Etc. Last group may contain less than 16384
+objects, and also contains the final key object.
+
+4. All objects within each group are sorted in according to their hashes.
+Objects are sorted by increasing of their hashes, precisely, by increasing
+of hex representations of hashes in lexicographic order. For example,
+the following is an example of sorted hashes in their hex representation:
+```
+0000000000000000000000000000000000000000000000000000000000000000
+154F29A919B30F50443A241C466691B046677C923EE7905AB97A4DBE8A5C2423
+2231553FC01D37A66C61BBEEACBB8C460994493E5659D118E19A8DDBB1444273
+272DCBFD8E4D5D786CF11A5444B30FB35435933B5DE6C660AA46E68CF0F5C447
+3C062FD9F0BCDCA31ACEBCD8E530D0BDAD1F1D1257B89C435616506A3EE6CB9E
+58A0E5AE427CDDC1C7C06448E8C3E4BF718DE036D827881624B20465C3E1334F
```

Comment:
Note that with thie example data we'd get the same sort order if we were sorting from the low order bits or the high order bits. It would be better to choose example data that sorted differently.

Reply:
Done.